File: 0346-moving-average-from-data-stream/0346-moving-average-from-data-stream.py
class MovingAverage:

    def __init__(self, size: int):
        self.size = size
        self.sum = 0
        self.arr = collections.deque([])
        self.count = 0
        

    def next(self, val: int) -> float:
        self.count += 1
        self.arr.append(val)
        self.sum += val
        if self.count > self.size: 
            leftval = self.arr[0]
            self.sum -= leftval
            self.arr.popleft()
        
        return self.sum / min(self.count, self.size)
        


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# param_1 = obj.next(val)

File: 0346-moving-average-from-data-stream/0346-moving-average-from-data-stream.js
/**
 * @param {number} size
 */
var MovingAverage = function(size) {
    this.size = size;
    this.count = 0;
    this.l = 0;
    this.arr = [];
    this.sum = 0;
};

/** 
 * @param {number} val
 * @return {number}
 */
MovingAverage.prototype.next = function(val) {
    this.count++;
    this.arr.push(val);
    this.sum += val;
    
    if (this.count > this.size) {
        let leftVal = this.arr[this.l];
        this.sum -= leftVal;
        this.l++;
    }
    return this.sum / Math.min(this.count, this.size);
    
};

/** 
 * Your MovingAverage object will be instantiated and called as such:
 * var obj = new MovingAverage(size)
 * var param_1 = obj.next(val)
 */